--- pkitool.bck	2015-11-28 14:54:58.303648598 +0100
+++ pkitool	2015-11-28 16:10:37.988593343 +0100
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
 
 #  OpenVPN -- An application to securely tunnel IP networks
 #             over a single TCP/UDP port, with support for SSL/TLS-based
@@ -32,7 +32,7 @@
 
 PROGNAME=pkitool
 VERSION=2.0
-DEBUG=0
+DEBUG=1
 
 die()
 {
@@ -134,6 +134,23 @@
 EOM
 }
 
+valid_ip()
+{
+    local  ip=$1
+    local  stat=1
+
+    if [[ $ip =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]]; then
+        OIFS=$IFS
+        IFS='.'
+        ip=($ip)
+        IFS=$OIFS
+        [[ ${ip[0]} -le 255 && ${ip[1]} -le 255 \
+            && ${ip[2]} -le 255 && ${ip[3]} -le 255 ]]
+        stat=$?
+    fi
+    return $stat
+}
+
 # Set tool defaults
 [ -n "$OPENSSL" ] || export OPENSSL="openssl"
 [ -n "$PKCS11TOOL" ] || export PKCS11TOOL="pkcs11-tool"
@@ -288,20 +305,19 @@
         FN="$1"
     fi
 else
-    if [ $# -ne 1 ]; then
-        usage
-        exit 1
-    else
         KEY_CN="$1"
+	FN="$KEY_CN"
 	KEY_ALTNAMES="DNS:$1"
 	shift
 	while [ "x$1" != "x" ]
 	do
-	    KEY_ALTNAMES="${KEY_ALTNAMES},DNS:$1"
+	    if valid_ip $1; then
+		KEY_ALTNAMES="${KEY_ALTNAMES},IP:$1"
+	    else
+		KEY_ALTNAMES="${KEY_ALTNAMES},DNS:$1"
+	    fi
 	    shift
 	done
-    fi
-    FN="$KEY_CN"
 fi
 
 export CA_EXPIRE KEY_EXPIRE KEY_OU KEY_NAME KEY_CN PKCS11_MODULE_PATH PKCS11_PIN KEY_ALTNAMES
